packages/editor/src/Editor.tsx

```diff
@@ -582,675 +599,790 @@ function calculateViewportRelatedStuff(
+    () =>
+      new ResizeObserver(
+        throttle(() => {
+          forceRerender();
+        }, 100)
+      ),
+    [forceRerender]
   );
 
   useEffect(() => {
     if (!iframe) {
-      return;
+      return () => {};
     }
 
-    resizeObserver.current.observe(iframe);
+    resizeObserver.observe(iframe);
 
     return () => {
-      resizeObserver.current.unobserve(iframe);
+      resizeObserver.unobserve(iframe);
     };
-  }, [iframe]);
+  }, [iframe, resizeObserver]);
 }
 
-const EditorContent = ({
-  compilationContext,
-  heightMode = "viewport",
-  initialDocument,
-  initialEntry,
-  externalData,
-  ...props
-}: EditorContentProps) => {
-  const [currentViewport, setCurrentViewport] = useState<string>(
-    compilationContext.mainBreakpointIndex
-  ); // "{ breakpoint }" or "fit-screen"
-
-  const iframeContainerRef = useRef<HTMLIFrameElement>(null);
-  const availableSize = iframeContainerRef.current
-    ? {
-        width: iframeContainerRef.current.clientWidth,
-        height: iframeContainerRef.current.clientHeight,
-      }
-    : undefined;
-
-  const { breakpointIndex, iframeSize } = calculateViewportRelatedStuff(
-    currentViewport,
-    compilationContext.devices,
-    compilationContext.mainBreakpointIndex,
-    availableSize
-  );
-
-  useRerenderOnIframeResize(iframeContainerRef.current); // re-render on resize (recalculates viewport size, active breakpoint for fit-screen etc);
-
-  const compilationCache = useRef(new CompilationCache());
-  const [isEditing, setEditing] = useState(true);
-  const [componentPickerData, setComponentPickerData] = useState<
-    | {
-        promiseResolve: (config: NoCodeComponentEntry | undefined) => void;
-        config: OpenComponentPickerConfig;
+const EditorContent = memo(
+  ({
+    compilationContext,
+    heightMode = "viewport",
+    initialDocument,
+    initialEntry,
+    externalData,
+    ...props
+  }: EditorContentProps) => {
+    const [currentViewport, setCurrentViewport] = useState<ViewportValue>(
+      () => {
+        return {
+          value: compilationContext.mainBreakpointIndex,
+          options: {},
+        };
       }
-    | undefined
-  >(undefined);
-  const [focussedField, setFocussedField] = useState<Array<string>>([]);
-
-  const handleSetFocussedField = React.useRef(
-    (field: Array<string> | string) => {
-      const nextFocusedField = Array.isArray(field) ? field : [field];
-      setFocussedField(nextFocusedField);
-    }
-  ).current;
+    ); // "{ breakpoint }" or "fit-screen"
+
+    const onViewportChange = useCallback((viewport: string) => {
+      const [value, options = ""] = viewport.split(":");
+
+      setCurrentViewport({
+        value,
+        options: options
+          .split(",")
+          .reduce((acc: Record<string, string>, option) => {
+            const [key, value] = option.split("=");
+            acc[key] = value;
+            return acc;
+          }, {}),
+      });
+    }, []);
```

Comment:
Now viewport has a different value format. In addition to the value, there are now options.

Example of a viewport value with options: `xs:height=max,margin=24`